src/codegeneration/SuperTransformer.js

```diff
@@ -12,159 +12,298 @@
+ *     }
+ *   }
  */
-export class SuperTransformer extends ParseTreeTransformer {
-  /**
-   * @param {TempVarTransformer} tempVarTransformer
-   * @param {ParseTree} protoName
-   * @param {string} thisName The name of the saved 'this' var
-   * @param {ParseTree} internalName The name of the save class binding.
-   */
-  constructor(tempVarTransformer, protoName, thisName, internalName) {
-    super();
-    this.tempVarTransformer_ = tempVarTransformer;
-    this.protoName_ = protoName;
-    this.internalName_ = internalName;
-    this.superCount_ = 0;
-    this.thisVar_ = createIdentifierExpression(thisName);
-    this.inNestedFunc_ = 0;
-    this.nestedSuperCount_ = 0;
+export class SuperTransformer extends TempVarTransformer {
+  constructor(identifierGenerator, reporter, options) {
+    super(identifierGenerator, reporter, options);
+    this.stateStack_ = [];
   }
 
-  get hasSuper() {
-    return this.superCount_ > 0;
+  pushState(state) {
+    this.stateStack_.push(state);
   }
 
-  get nestedSuper() {
-    return this.nestedSuperCount_ > 0;
+  popState() {
+    return this.stateStack_.pop();
   }
 
-  transformFunctionDeclaration(tree) {
-    return this.transformFunction_(tree, FunctionDeclaration);
+  peekState() {
+    return this.stateStack_[this.stateStack_.length - 1];
   }
 
-  transformFunctionExpression(tree) {
-    return this.transformFunction_(tree, FunctionExpression);
+  transformObjectLiteralExpression(tree) {
+    let state = new State(this, null);
+    this.pushState(state);
+    this.pushState(state);
+    let result = super.transformObjectLiteralExpression(tree);
+    this.popState();
+    this.popState();
+    if (state.hasSuper) {
+      this.registerTempVarName(state.tempName);
+      return createAssignmentExpression(state.home, result);
+    }
+    this.releaseTempName(state.tempName);
+    return result;
   }
 
-  transformFunction_(tree, constructor) {
-    let oldSuperCount = this.superCount_;
+  transformClassExpression(tree) {
+    let superClass = this.transformAny(tree.superClass);
+    let annotations = this.transformList(tree.annotations);
+
+    let classState = new ClassState(this, tree);
+    let prototypeState = new PrototypeState(this, classState);
 
-    this.inNestedFunc_++;
-    let transformedTree = constructor === FunctionExpression ?
-        super.transformFunctionExpression(tree) :
-        super.transformFunctionDeclaration(tree);
-    this.inNestedFunc_--;
+    this.pushState(classState);
+    this.pushState(prototypeState);
+    let elements = this.transformList(tree.elements);
+    this.popState();
+    this.popState();
 
-    if (oldSuperCount !== this.superCount_)
-      this.nestedSuperCount_ += this.superCount_ - oldSuperCount;
+    if (tree.name === null && tree.superClass !== null) {
+      // In case the class expression has no constructor and it has an extends
+      // clause use force a name to be added.
```

Comment:
I don't understand `clause use force a name to be added.`  Maybe "call the home accessor for its side-effect"?